drivers/fpga/dfl-pci-sva.c

```diff
@@ -0,0 +1,324 @@
+					       struct dfl_sva_dev, mdev_fops);
+	struct iommu_sva **sva_handle_p;
+	long ret;
+
+	sva_handle_p = (struct iommu_sva **)&file->private_data;
+
+	switch (cmd) {
+	case DFL_FPGA_GET_API_VERSION:
+		ret = DFL_FPGA_API_VERSION;
+		break;
+	case DFL_PCI_SVA_BIND_DEV:
+		ret = ioctl_sva_bind_dev(dev, sva_handle_p);
+		break;
+	case DFL_PCI_SVA_UNBIND_DEV:
+		ret = ioctl_sva_unbind_dev(dev, sva_handle_p);
+		break;
+	default:
+		pci_info(dev->pdev, "%x cmd not handled", cmd);
+		ret = -EINVAL;
+	}
+
+	return ret;
+}
+
+static int add_dfl_mdev(struct dfl_sva_dev *dev)
+{
+	struct pci_dev *pdev = dev->pdev;
+
+	snprintf(dev->mdev_name, sizeof(dev->mdev_name),
+		 "dfl-pci-sva!%04x:%02x:%02x.%x",
+		 pci_domain_nr(pdev->bus),
+		 pdev->bus->number,
+		 PCI_SLOT(pdev->devfn),
+		 PCI_FUNC(pdev->devfn));
+
+	dev->mdev_fops.open = dfl_pci_sva_open;
+	dev->mdev_fops.release = dfl_pci_sva_release;
+	dev->mdev_fops.unlocked_ioctl = dfl_pci_sva_ioctl;
+	dev->mdev_fops.owner = THIS_MODULE;
+
+	dev->mdev.minor = MISC_DYNAMIC_MINOR;
+	dev->mdev.name = dev->mdev_name;
+	dev->mdev.fops = &dev->mdev_fops;
+	dev->mdev.mode = 0400;
+
+	misc_register(&dev->mdev);
+
+	return 0;
+}
+
+static void del_dfl_mdev(struct dfl_sva_dev *dev)
+{
+	pci_info(dev->pdev, "dfl-sva delete device\n");
+	misc_deregister(&dev->mdev);
+	disable_iommu_sva_feature(dev->pdev);
+}
+
+static inline bool is_dfl_device(struct pci_dev *pdev)
+{
+	return pdev->vendor == PCI_VENDOR_ID_INTEL &&
+	       (pdev->device == PCIE_DEVICE_ID_INTEL_DFL ||
+		pdev->device == PCIE_DEVICE_ID_INTEL_DFL_VF);
+}
+
+/*
+ * Consider adding a new device. This is called both by new dfl-pci probes
+ * and from the PCIe bus notifier.
+ */
+void dfl_pci_sva_add_dev(struct pci_dev *pdev)
+{
+	int ret;
+	struct dfl_sva_dev *cur;
+
+	/*
+	 * Consider new DFL and DFL_VF devices, adding them to dfl_dev_list
+	 * if they support shared virtual addressing.
+	 */
+	if (is_dfl_device(pdev)) {
+		mutex_lock(&dfl_dev_list_mutex);
+		/* Nothing to do if the device was already added */
+		list_for_each_entry(cur, &dfl_dev_list, pdev_next) {
+			if (cur->pdev == pdev)
+				goto out_unlock;
+		}
+
+		/* Manage only devices with SVA features (returns 0 here) */
+		if (enable_iommu_sva_feature(pdev))
+			goto out_unlock;
+
+		cur = kzalloc(sizeof(struct dfl_sva_dev), GFP_KERNEL);
+		if (!cur)
+			goto out_disable;
+
+		pci_info(pdev, "dfl-sva add device\n");
+		cur->pdev = pdev;
+
+		ret = add_dfl_mdev(cur);
+		if (ret) {
+			kfree(cur);
+			goto out_disable;
+		}
+
+		list_add(&cur->pdev_next, &dfl_dev_list);
+		mutex_unlock(&dfl_dev_list_mutex);
+	}
+
+	return;
+
+out_disable:
+	disable_iommu_sva_feature(pdev);
+out_unlock:
+	mutex_unlock(&dfl_dev_list_mutex);
+}
+
+static void dfl_pci_sva_del_dev(struct pci_dev *pdev)
+{
+	struct dfl_sva_dev *cur;
+
+	if (is_dfl_device(pdev)) {
+		mutex_lock(&dfl_dev_list_mutex);
+		list_for_each_entry(cur, &dfl_dev_list, pdev_next) {
+			if (cur->pdev == pdev) {
```

Comment:
I don't think I need "safe" here because traversal stops as soon as the entry to delete is found. Safe just makes a copy of the thing I'm deleting in order to find next, right?